packages/performance/src/services/transport_service.ts

```diff
@@ -73,60 +82,121 @@ function processQueue(timeOffset: number): void {
-
-    const data: CcBatchLogFormat = {
-      request_time_ms: String(Date.now()),
-      client_info: {
-        client_type: 1, // 1 is JS
-        js_client_info: {}
-      },
-      log_source: SettingsService.getInstance().logSource,
-      log_event
-    };
-    /* eslint-enable camelcase */
-
-    fetch(SettingsService.getInstance().logEndPointUrl, {
-      method: 'POST',
-      body: JSON.stringify(data)
+    dispatchQueueEvents();
+  }, timeOffset);
+}
+
+function dispatchQueueEvents(): void {
+  // Capture a snapshot of the queue and empty the "official queue".
+  const staged = [...queue];
+  queue = [];
+
+  /* eslint-disable camelcase */
+  // We will pass the JSON serialized event to the backend.
+  const log_event: Log[] = staged.map(evt => ({
+    source_extension_json_proto3: evt.message,
+    event_time_ms: String(evt.eventTime)
+  }));
+
+  const data: TransportBatchLogFormat = {
+    request_time_ms: String(Date.now()),
+    client_info: {
+      client_type: 1, // 1 is JS
+      js_client_info: {}
+    },
+    log_source: SettingsService.getInstance().logSource,
+    log_event
+  };
+  /* eslint-enable camelcase */
+
+  postToEndpoint(data, staged).catch(() => {
+    // If the request fails for some reason, add the events that were attempted
+    // back to the primary queue to retry later.
+    queue = [...staged, ...queue];
+    remainingTries--;
+    consoleLogger.info(`Tries left: ${remainingTries}.`);
+    processQueue(DEFAULT_SEND_INTERVAL_MS);
+  });
+}
+
+function postToEndpoint(
+  data: TransportBatchLogFormat,
+  staged: BatchEvent[]
+): Promise<void> {
+  // Gradually rollout traffic from cc to transport using remote config.
+  if (SettingsService.getInstance().shouldSendToTransport) {
+    return sendEventsToTransport(data, staged);
+  } else {
+    return sendEventsToCc(data);
+  }
+}
+
+function sendEventsToTransport(
+  data: TransportBatchLogFormat,
+  staged: BatchEvent[]
+): Promise<void> {
+  return postToTransportEndpoint(data)
+    .then(res => {
+      if (!res.ok) {
+        consoleLogger.info('Call to Firebase backend failed.');
+      }
+      return res.json();
     })
-      .then(res => {
-        if (!res.ok) {
-          consoleLogger.info('Call to Firebase backend failed.');
+    .then(res => {
+      // Find the next call wait time from the response.
+      const transportWait = Number(res.nextRequestWaitMillis);
+      let requestOffset = DEFAULT_SEND_INTERVAL_MS;
+      if (!isNaN(transportWait)) {
+        requestOffset = Math.max(transportWait, requestOffset);
+      }
+
+      // Delete request if response include RESPONSE_ACTION_UNKNOWN or DELETE_REQUEST action.
+      // Otherwise, retry request using normal scheduling if response include RETRY_REQUEST_LATER.
+      const logResponseDetails: LogResponseDetails[] = res.logResponseDetails;
+      if (logResponseDetails && logResponseDetails.length > 0) {
+        if (logResponseDetails[0].responseAction === 'RETRY_REQUEST_LATER') {
```

Comment:
We need to check for logResponseDetails to be an array before checking the length. Also, the two if statement can merge.